src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1031,325 @@ begin
 end
 
 end matrix_adjoints
+
+namespace bilin_form
+
+section orthogonal
+
+/-- The orthogonal complement of a submodule `N` with respect to some bilinear form is the set
+  of elements which are orthogonal to all elements of `N`. -/
+def orthogonal (B : bilin_form R M) (N : submodule R M) : submodule R M :=
+{ carrier := { m | ∀ n ∈ N, is_ortho B m n },
+  zero_mem' := λ x _, is_ortho_zero_left x,
+  add_mem' := λ x y hx hy n hn,
+    by rw [is_ortho, add_left, show B x n = 0, by exact hx n hn,
+        show B y n = 0, by exact hy n hn, zero_add],
+  smul_mem' := λ c x hx n hn,
+    by rw [is_ortho, smul_left, show B x n = 0, by exact hx n hn, mul_zero] }
+
+variables {N L : submodule R M}
+
+@[simp] lemma mem_ortho_iff {N : submodule R M} {m : M} :
+  m ∈ B.orthogonal N ↔ ∀ n ∈ N, is_ortho B m n := iff.rfl
+
+lemma orthogonal_le (h : N ≤ L) : B.orthogonal L ≤ B.orthogonal N :=
+  λ _ hn l hl, hn l (h hl)
+
+lemma le_orthogonal_orthogonal (hB : refl_bilin_form.is_refl B) :
+  N ≤ B.orthogonal (B.orthogonal N) := λ n hn m hm, hB _ _ (hm n hn)
+
+end orthogonal
+
+/-- A set of vectors `v` is orthogonal with respect to some bilinear form `B` if and only if for
+  all `i ≠ j`, `B (v i) (v j) = 0`. -/
+def is_ortho' {n : Type w} (B : bilin_form R M) (v : n → M) : Prop :=
+  ∀ i j : n, i ≠ j → B (v j) (v i) = 0
```

Comment:
The link to `is_ortho` might be more obvious as
```suggestion
  ∀ i j : n, i ≠ j → is_ortho (v j) (v i)
```

Reply:
I agree with this, but it does break some stuff which I'll fix in the next commit